nemo/collections/nlp/data/text_normalization_as_tagging/utils.py

```diff
@@ -0,0 +1,300 @@
+
+    # remove same fragments at the beginning or at the end of spoken and written form
+    written_parts = written.split()
+    spoken_parts = spoken.split()
+    same_from_begin = 0
+    same_from_end = 0
+    for i in range(min(len(written_parts), len(spoken_parts))):
+        if written_parts[i] == spoken_parts[i]:
+            same_from_begin += 1
+        else:
+            break
+    for i in range(min(len(written_parts), len(spoken_parts))):
+        if written_parts[-i - 1] == spoken_parts[-i - 1]:
+            same_from_end += 1
+        else:
+            break
+    same_begin = written_parts[0:same_from_begin]
+    same_end = []
+    if same_from_end == 0:
+        written = " ".join(written_parts[same_from_begin:])
+        spoken = " ".join(spoken_parts[same_from_begin:])
+    else:
+        written = " ".join(written_parts[same_from_begin:-same_from_end])
+        spoken = " ".join(spoken_parts[same_from_begin:-same_from_end])
+        same_end = written_parts[-same_from_end:]
+
+    fragments = list(split_text_by_isalpha(written))
+    written_tokens = []
+    for frag in fragments:
+        if frag.isalpha():
+            if semiotic_class == "plain" or semiotic_class == "letters" or semiotic_class == "electronic":
+                chars = list(frag.strip())
+                chars[0] = "_" + chars[0]  # prepend first symbol of a word with underscore
+                chars[-1] = chars[-1] + "_"  # append underscore to the last symbol
+                written_tokens += chars
+            else:
+                written_tokens.append("_" + frag + "_")
+        else:
+            chars = list(frag.strip().replace(" ", ""))
+            if len(chars) > 0:
+                chars[0] = "_" + chars[0]  # prepend first symbol of a non-alpha fragment with underscore
+                chars[-1] = chars[-1] + "_"  # append underscore to the last symbol of a non-alpha fragment
+                written_tokens += chars
+    written_str = " ".join(written_tokens)
+
+    # _н_ _._ _г_ _._ => _н._ _г._
+    written_str = re.sub(
+        r"([abcdefghijklmnopqrstuvwxyzабвгдеёжзийклмнопрстуфхцчшщъыьэюя])_ _\._", r"\g<1>._", written_str
+    )
+    # _тыс_ _. $ => _тыс._ _$
+    written_str = re.sub(
+        r"([abcdefghijklmnopqrstuvwxyzабвгдеёжзийклмнопрстуфхцчшщъыьэюя])_ _\. ([^_])]", r"\g<1>._ _\g<2>", written_str
+    )
+
+    if semiotic_class == "ordinal":
+        #  _8 2 -_ _ом_  =>  _8 2-ом_
+        written_str = re.sub(
+            r"([\d]) -_ _([abcdefghijklmnopqrstuvwxyzабвгдеёжзийклмнопрстуфхцчшщъыьэюя]+)_",
+            r"\g<1>-\g<2>_",
+            written_str,
+        )
+        #  _8 8_ _й_       _8 8й_
+        written_str = re.sub(
+            r"([\d])_ _([abcdefghijklmnopqrstuvwxyzабвгдеёжзийклмнопрстуфхцчшщъыьэюя]+)_", r"\g<1>\g<2>_", written_str
+        )
+
+    if semiotic_class == "cardinal":
+        #  _2 5 -_ _ти_ => _2 5-ти_
+        written_str = re.sub(r"([\d]) -_ _(ти)_", r"\g<1>-\g<2>_", written_str)
```

Comment:
this is post corrections after alignment but before the training?

Reply:
no, this is some preprocessing **before** alignment, to simplify it for Giza

Reply:
move to pre-processing as discussed

Reply:
done